File: WoofAndroid/konsist/src/test/kotlin/com/perrystreet/woof/konsist/viewmodel/ViewModelsLiveInViewModelPackage.kt
package com.perrystreet.woof.konsist.viewmodel

import com.perrystreet.woof.konsist.Assertions.assertTrue
import com.perrystreet.woof.konsist.KonsistUtils
import com.perrystreet.woof.konsist.LintRuleMessage
import io.kotest.core.spec.style.BehaviorSpec

class ViewModelsLiveInViewModelPackage : BehaviorSpec() {
    init {
        Given("A ViewModel") {
            val viewModels = KonsistUtils.viewModels

            Then("It is declared in the feature's viewmodel package") {
                viewModels.assertTrue(message = Message) { it.resideInPackage("com.perrystreet.woof.presentation..viewmodel") }
            }
        }
    }

    private companion object {
        private val Message = LintRuleMessage(
            rule = "ViewModels live in presentation/<feature>/viewmodel.",
            why = "Each feature module has the same four folders (viewmodel, uimodel, mapper, ui), so anyone can find a class by convention.",
            howToFix = "Move the class to the viewmodel package of its feature.",
            badExample = "package com.perrystreet.woof.presentation.grid.ui\nclass GridViewModel",
            goodExample = "package com.perrystreet.woof.presentation.grid.viewmodel\nclass GridViewModel",
        )
    }
}

File: WoofIOS/SwiftPackages/HarmonizeRules/Tests/HarmonizeRulesTests/ViewModel/ViewModelsLiveInViewModelPackage.swift
import Foundation
import Harmonize
import HarmonizeSemantics
import Quick

final class ViewModelsLiveInViewModelPackage: QuickSpec {
    override class func spec() {
        Given("A ViewModel") {
            let viewModels = WoofHarmonize.viewModels

            Then("It is declared in the feature's ViewModel folder") {
                viewModels.assertTrue(rule: rule) { $0.filePathString.contains("/ViewModel/") }
            }
        }
    }

    private static let rule = Rule(
        description: "ViewModels live in Presentation/<Feature>/Sources/<Target>/ViewModel.",
        rationale: "Each feature package has the same four folders (ViewModel, UIModel, Mapper, UI), so anyone can find a class by convention.",
        fixHint: "Move the class to the ViewModel folder of its feature.",
        badExample: "// Presentation/Grid/Sources/PresentationGrid/UI/GridViewModel.swift",
        goodExample: "// Presentation/Grid/Sources/PresentationGrid/ViewModel/GridViewModel.swift"
    )
}
